packages/xo-web/src/xo-app/sr/tab-disks.js

```diff
@@ -290,10 +307,83 @@ class NewDisk extends Component {
   }
 }
 
-@connectStore(() => ({
-  checkPermissions: getCheckPermissions,
-  vbds: createGetObjectsOfType('VBD'),
-}))
+@connectStore(() => {
+  const getAllVdis = createGetObjectsOfType('VDI')
+  const getSrs = createGetObjectsOfType('SR')
+  const getVbds = createGetObjectsOfType('VBD')
+
+  const getVdisById = createSelector(
+    (_, props) => props.vdis,
+    vdis => keyBy(vdis, 'id')
+  )
+
+  const getVbdsByVdi = createSelector(
+    createSelector(getVbds, vbds => keyBy(vbds, 'id')),
+    createSelector(getVdisById, vdis => mapValues(vdis, '$VBDs')),
+    (vbds, vbdIdsByVdi) =>
+      mapValues(vbdIdsByVdi, vbdIds => map(vbdIds, vbdId => vbds[vbdId]))
+  )
+
+  const getVmIdsByVdi = createSelector(getVbdsByVdi, vbdsByVdi =>
+    mapValues(vbdsByVdi, vbds => map(vbds, 'VM'))
+  )
+
+  const getVmsByVdi = createSelector(
+    createGetObjectsOfType('VM'),
+    getVmIdsByVdi,
+    (vms, vmIdsByVdi) => mapValues(vmIdsByVdi, vmIds => pick(vms, vmIds))
+  )
+
+  const getVmControllersByVdi = createSelector(
+    createGetObjectsOfType('VM-controller'),
+    getVmIdsByVdi,
+    (vmControllers, vmIdsByVdi) =>
+      mapValues(vmIdsByVdi, vmIds => pick(vmControllers, vmIds))
+  )
+
+  const getVmSnapshotsByVdi = createSelector(
+    createGetObjectsOfType('VM-snapshot'),
+    getVmIdsByVdi,
+    (vmSnapshots, vmIdsByVdi) =>
+      mapValues(vmIdsByVdi, vmIds => pick(vmSnapshots, vmIds))
+  )
+
+  const getVmTemplatesByVdi = createSelector(
+    createGetObjectsOfType('VM-template'),
+    getVmIdsByVdi,
+    (vmTemplates, vmIdsByVdi) =>
+      mapValues(vmIdsByVdi, vmIds => pick(vmTemplates, vmIds))
+  )
+
+  const getAllVmsByVdi = createSelector(
```

Comment:
To avoid so mush indirection, i suggest to implement a selector which takes a `VDI` and returns a callback to check the  `SR`.